File: LeetCode/Binary Search/Easy/704. Binary Search/main.cpp
/*
Binary Search -  we recursively divide the array in half to find the element
*/

#include <vector>
#include <iostream>

using namespace std;
class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        int low = 0;
        int high = nums.size() - 1;
        while (low <= high)
        {
            int mid = low + (high - low + 1) / 2;
            cout << low << mid << high << endl;
            if (nums[mid] == target)
                return mid;
            else if (nums[mid] < target)
                low = mid + 1;
            else
                high = mid - 1;
            cout << low << mid << high << endl;
        }
        return -1;
    }
};

int main()
{
    Solution x;
    vector<int> y{-1, 0, 3, 5, 9, 12};
    int out{x.search(y, 9)};
    cout << out;
    return 0;
}

File: LeetCode/Binary Search/Easy/704. Binary Search/main.py
"""
Binary Search -  we recursively divide the array in half to find the element 
"""


class Solution:
    def search(self, nums, target):
        # print(nums,target)
        low = 0
        high = len(nums)-1
        while(low <= high):
            mid = low + (high-low+1)//2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                high = mid - 1               # mid is excluded
            else:
                low = mid + 1                   # mid is included
            # print(high, low, mid)
        return -1
        return mid # if we want to return where element can be pushed in array


x = Solution()
print(x.search([-1, 0, 3, 5, 9, 12], 9))
